src/cmd/builtin/fsck.c

```diff
@@ -0,0 +1,266 @@
+        log_err_exit ("%s: could not decode directory", path);
+    if (!treeobj_is_dir (treeobj_deref))
+        log_msg_exit ("%s: dirref references non-directory", path);
+    fsck_dir (h, path, treeobj_deref); // recurse
+    json_decref (treeobj_deref);
+    flux_future_destroy (f);
+}
+
+static void fsck_treeobj (flux_t *h,
+                          const char *path,
+                          json_t *treeobj)
+{
+    if (treeobj_validate (treeobj) < 0)
+        log_msg_exit ("%s: invalid tree object", path);
+    if (treeobj_is_symlink (treeobj)) {
+        if (verbose)
+            fprintf (stderr, "%s\n", path);
+        fsck_symlink (h, path, treeobj);
+    }
+    else if (treeobj_is_val (treeobj)) {
+        if (verbose)
+            fprintf (stderr, "%s\n", path);
+        fsck_val (h, path, treeobj);
+    }
+    else if (treeobj_is_valref (treeobj)) {
+        if (verbose)
+            fprintf (stderr, "%s\n", path);
+        fsck_valref (h, path, treeobj);
+    }
+    else if (treeobj_is_dirref (treeobj)) {
+        fsck_dirref (h, path, treeobj); // recurse
+    }
+    else if (treeobj_is_dir (treeobj)) {
+        fsck_dir (h, path, treeobj); // recurse
+    }
+}
+
+static void fsck_blobref (flux_t *h, const char *blobref)
+{
+    flux_future_t *f;
+    const void *buf;
+    size_t buflen;
+    json_t *treeobj;
+    json_t *dict;
+    const char *key;
+    json_t *entry;
+
+    if (!(f = content_load_byblobref (h, blobref, CONTENT_FLAG_CACHE_BYPASS))
+        || content_load_get (f, &buf, &buflen) < 0) {
+        read_error ("cannot load root tree object: %s",
+                    future_strerror (f, errno));
+        errorcount++;
+        flux_future_destroy (f);
+        return;
+    }
+    if (!(treeobj = treeobj_decodeb (buf, buflen)))
+        log_err_exit ("cannot decode root tree object");
+    if (treeobj_validate (treeobj) < 0)
+        log_msg_exit ("invalid root tree object");
+    if (!treeobj_is_dir (treeobj))
+        log_msg_exit ("root tree object is not a directory");
+
+    dict = treeobj_get_data (treeobj);
+    json_object_foreach (dict, key, entry) {
+        fsck_treeobj (h, key, entry);
+    }
+    json_decref (treeobj);
+    flux_future_destroy (f);
+}
+
+static int cmd_fsck (optparse_t *p, int ac, char *av[])
+{
+    int optindex =  optparse_option_index (p);
+    flux_future_t *f;
+    const char *blobref;
+    flux_t *h;
+
+    log_init ("flux-fsck");
+
+    if (optindex != ac) {
+        optparse_print_usage (p);
+        exit (1);
+    }
+
+    if (optparse_hasopt (p, "verbose"))
+        verbose = true;
+
+    h = builtin_get_flux_handle (p);
+
+    if (!(f = kvs_checkpoint_lookup (h,
+                                     NULL,
+                                     KVS_CHECKPOINT_FLAG_CACHE_BYPASS))
+        || kvs_checkpoint_lookup_get_rootref (f, &blobref) < 0)
+        log_msg_exit ("error fetching checkpoint: %s",
+                      future_strerror (f, errno));
+
```

Comment:
It would be useful here to use `kvs_checkpoint_lookup_get_timetstamp()` and print something like
```
Checking integrity of stable checkpoint from <date>
```
Possibly a `--quiet` option would be useful to suppress this message.